go/kvstore/cache.go

```diff
@@ -42,45 +44,104 @@ func (e KVRevisionCacheError) Error() string {
 	return e.Message
 }
 
-func (k *KVRevisionCache) PutCheck(mctx libkb.MetaContext, entryID keybase1.KVEntryID, entryHash string, teamKeyGen keybase1.PerTeamKeyGeneration, revision int) (err error) {
+// Hash is a sha256 on the input string. If the string is empty, then Hash will also be an
+// empty string for tracking deleted entries in perpetuity.
+func (k *KVRevisionCache) hash(ciphertext *string) string {
+	if ciphertext == nil || len(*ciphertext) == 0 || *ciphertext == DELETED_OR_NONEXISTENT {
+		return DELETED_OR_NONEXISTENT
+	}
+	b := sha256.Sum256([]byte(*ciphertext))
+	return hex.EncodeToString(b[:])
+}
+
+func (k *KVRevisionCache) Check(mctx libkb.MetaContext, entryID keybase1.KVEntryID, ciphertext *string, teamKeyGen keybase1.PerTeamKeyGeneration, revision int) (err error) {
+	k.Lock()
+	defer k.Unlock()
+	k.ensureIntermediateLocked(entryID)
+
+	entry, ok := k.data[entryID.TeamID][entryID.Namespace][entryID.EntryKey]
+	if !ok {
+		// this entry didn't exist in the cache, so there's nothing to check
+		return nil
+	}
+	entryHash := k.hash(ciphertext)
+	if revision < entry.Revision {
+		return KVRevisionCacheError{fmt.Sprintf("cache error: revision decreased from %d to %d", entry.Revision, revision)}
+	}
+	if teamKeyGen < entry.TeamKeyGen {
+		return KVRevisionCacheError{fmt.Sprintf("cache error: team key generation decreased from %d to %d", entry.TeamKeyGen, teamKeyGen)}
+	}
+	if revision == entry.Revision {
+		if teamKeyGen != entry.TeamKeyGen {
+			return KVRevisionCacheError{fmt.Sprintf("cache error: at the same revision (%d) team key gen cannot be different: %d -> %d", revision, entry.TeamKeyGen, teamKeyGen)}
+		}
+		if entryHash != entry.EntryHash {
+			return KVRevisionCacheError{fmt.Sprintf("cache error: at the same revision (%d) hash of entry cannot be different: %s -> %s", revision, entry.EntryHash, entryHash)}
+		}
+	}
+	return nil
+}
+
+func (k *KVRevisionCache) Put(mctx libkb.MetaContext, entryID keybase1.KVEntryID, ciphertext *string, teamKeyGen keybase1.PerTeamKeyGeneration, revision int) (err error) {
 	k.Lock()
 	defer k.Unlock()
 
 	k.ensureIntermediateLocked(entryID)
+	entryHash := k.hash(ciphertext)
 	newEntry := kvCacheEntry{
 		EntryHash:  entryHash,
 		TeamKeyGen: teamKeyGen,
 		Revision:   revision,
 	}
+	k.data[entryID.TeamID][entryID.Namespace][entryID.EntryKey] = newEntry
+	return nil
+}
+
+func (k *KVRevisionCache) CheckDeletable(mctx libkb.MetaContext, entryID keybase1.KVEntryID, revision int) (err error) {
+	k.Lock()
+	defer k.Unlock()
+	k.ensureIntermediateLocked(entryID)
+
 	entry, ok := k.data[entryID.TeamID][entryID.Namespace][entryID.EntryKey]
-	if ok {
-		// the cache knows about this
-		err = checkNewAgainstCachedEntry(newEntry, entry)
-		if err != nil {
-			mctx.Debug("KVRevisionCache hit but with a mismatch from the server: %v", err)
-			return err
+	if !ok {
+		return KVRevisionCacheError{fmt.Sprintf("cannot delete an unknown entry")}
```

Comment:
huh. If the end user is going to get errors when deleting non-existent keys. There will need to be a way to recognize those errors separate from general errors. So that they can ensure a key does not exist.

Reply:
if you fetch an entry that doesn't exist, it'll have an empty string for a value and a revision of 0. does that resolve the concern?

Reply:
Because you can fetch to check whether it was deleted? That requires an extra rpc and seems like a workaround.

Reply:
yeah so actually, this should never execute. `ensureIntermediateLocked` right above ensures that there's at least an empty entry in there. so `ok` should never ever be false. i can change that error message to be more like a `"this should never happen"` style error. or we can panic and force a restart. i just figured it was better to do something than nothing.

Reply:
i'm gonna think about this. it doesn't seem like it's really working in an ideal way.